website/project/model.py

```diff
@@ -2829,188 +2898,302 @@ def to_json(self):
-            if parent_registration.embargo_end_date or parent_registration.pending_embargo:
-                parent_registration.embargo.state = self.CANCELLED
-                parent_registration.registered_from.add_log(
-                    action=NodeLog.EMBARGO_CANCELLED,
-                    params={
-                        'node': parent_registration._id,
-                        'embargo_id': parent_registration.embargo._id,
-                    },
-                    auth=Auth(user),
-                )
-                parent_registration.embargo.save()
-            # Ensure retracted registration is public
-            if not parent_registration.is_public:
-                parent_registration.set_privacy('public')
-            parent_registration.update_search()
-            # Retraction status is inherited from the root project, so we
-            # need to recursively update search for every descendant node
-            # so that retracted subrojects/components don't appear in search
-            for node in parent_registration.get_descendants_recursive():
-                node.update_search()
-
-
-def validate_embargo_state(value):
-    acceptable_states = [
-        Embargo.UNAPPROVED, Embargo.ACTIVE, Embargo.CANCELLED, Embargo.COMPLETED
-    ]
-    if value not in acceptable_states:
-        raise ValidationValueError('Invalid embargo state assignment.')
-    return True
+    def reject(self, user, token):
+        """Cancels sanction if user is admin and token verifies."""
+        try:
+            if self.approval_state[user._id]['rejection_token'] != token:
+                raise InvalidSanctionRejectionToken(self.REJECTION_INVALID_TOKEN_MESSAGE.format(DISPLAY_NAME=self.DISPLAY_NAME))
+        except KeyError:
+            raise PermissionsError(self.REJECTION_NOT_AUTHORIZED_MESSAEGE.format(DISPLAY_NAME=self.DISPLAY_NAME))
+        self.state = Sanction.REJECTED
+        self._on_reject(user, token)
 
+    def _notify_authorizer(self, user):
+        pass
 
-class Embargo(StoredObject):
-    """Embargo object for registrations waiting to go public."""
+    def _notify_non_authorizer(self, user):
+        pass
 
-    UNAPPROVED = 'unapproved'
-    ACTIVE = 'active'
-    CANCELLED = 'cancelled'
-    COMPLETED = 'completed'
+    def ask(self, group):
+        for contrib in group:
+            if contrib._id in self.approval_state:
+                self._notify_authorizer(contrib)
+            else:
+                self._notify_non_authorizer(contrib)
 
-    _id = fields.StringField(primary=True, default=lambda: str(ObjectId()))
-    initiation_date = fields.DateTimeField(auto_now_add=datetime.datetime.utcnow)
-    initiated_by = fields.ForeignField('user', backref='embargoed')
-    end_date = fields.DateTimeField()
-    # Expanded: Dictionary field mapping admin IDs their approval status and relevant tokens:
+class EmailApprovableSanction(Sanction):
+
+    AUTHORIZER_NOTIFY_TEMPLATE = None
```

Comment:
These need to be `Mail` objects, correct? Perhaps rename this variable to reflect that and/or document it.